src/librustc/middle/region.rs

```diff
@@ -1166,6 +1168,97 @@ fn region_maps<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId)
     Rc::new(maps)
 }
 
+struct YieldFinder<'a> {
+    cache: &'a mut FxHashMap<NodeId, Option<Span>>,
+    result: Option<Span>,
+}
+
+impl<'a> YieldFinder<'a> {
+    fn lookup<F: FnOnce(&mut Self)>(&mut self, id: NodeId, f: F) {
+        if let Some(result) = self.cache.get(&id) {
+            self.result = *result;
+            return;
+        }
+        if self.result.is_some() {
```

Comment:
This is subtly correct and probably warrants a comment somewhere: this populates all subtrees with `None` until it find a `Yield`, and then it marks the upwards branch from that `Yield` with `Some(span)` *and does not visit any other subtree*.

```Rust
// When we go
```